[flang][OpenMP] Re-enable tests when building OpenMP as a runtime

mjklemm opened the pull request:
This PR re-enables OpenMP tests when the OpenMP library is configured as a runtime build.  After PR #80874 has disabled them for a while.

The test system now has a command substitution that adds "-J" with the proper folder that contains omp_lib.{h,mod} to the Flang compiler command line in a test.  This is only done when the OpenMP library is a runtime build (via `LLVM_ENABLE_RUNTIMES`), but is skipped if the OpenMP library is built as a regular LLVM project (via `LLVM_ENABLE_PROJECTS`)

mjklemm committed:
Add command substitution to conditionally add proper -J flag to test

mjklemm committed:
Add %openmp_module_flag to Flang invocation

mjklemm commented:
This PR cannot land before PR #88934 has landed.

mjklemm committed:
Make darker happy

banach-space commented:
Why not update substitutions for `%flang_fc1`/`%bbc` etc? This way you could avoid updating so many tests. And RUN lines would be shorter.

I don't have a strong opinion. Thinking out loud - the proposed implementation that even people who don't work on OpenMP will have to remember to add this extra variable to RUN lines. Admittedly, updating %flang_fc1`/`%bbc` obfuscates things a bit. But that would only affect OpenMP experts?

mjklemm commented:
I was playing with the option to extend the `extra_args` field of `ToolSubst`.  But then, if the OpenMP runtime is enabled, it would inject `-J` for all tests not only the OpenMP tests.  I kind of felt that this was a bit too much.  That's why I went with the option of having `%openmp_module_flag` and putting the burden on the OpenMP tests only.  I was hoping that the usual test of creating a new test is by "borrowing" from an existing test and thus automatically getting `%openmp_module_flag` into the new test.

Other options that I see (happy to get more)
- Have an extension to Flang to pick up `-J` and `-I` from the environment (similar to `LIBRARY_PATH` or `INCLUDE_PATH` with C compilers)
- Change the build system to copy/link/move omp_lib.mod from where it lives for runtime builds to where it lives when doing a project build

mjklemm commented:
> Just a suggestion: we can combine `-J ...` and `-fopenmp`, and name it just `%openmp_flag`. This will still require changes in all tests, but going forward we can add more things in `%openmp_flag`, if needed, and all OpenMP tests will use it instead of `-fopenmp`.

I kind of like that!  I'll craft an update tomorrow and post it here.  Thanks for the suggestion!

mjklemm committed:
Move -fopenmp into %openmp_flags together with -J (if used)

tblah reviewed (approved):
Looks good to me. Thanks for the update

banach-space reviewed (approved):
I missed that this change only impacts OpenMP tests. This LGTM, thanks!

mjklemm commented:
Thanks all for the approval.  I will let this PR sit until the dependent PR has been merged.  Not sure how long that will take.  If it will take too long, I will ask @kparzysz to re-open #88914 and land it.

mjklemm committed:
Detect runtime build of the OpenMP runtime

Inspired by https://github.com/llvm/llvm-project/pull/88914

kiranchandramohan commented:
All the FIR-based tests (eg flang/test/Lower/OpenMP/FIR/teams.f90) are now removed. You might need a rebase.

mjklemm committed:
Resolve conflict

mjklemm commented:
@kiranchandramohan @banach-space I think this go in now.  If you agree, please approve, that I can merge.  The conflicts seem easy to resolve, so I'd do this during merge.

kiranchandramohan reviewed (approved):
LGTM.

mjklemm committed:
Set two tests to XFAIL as they will be failing when merged

These tests will fail in the main branch when this PR is being merged.  This is not related to PR.

mjklemm committed:
Merge branch 'main' into openmp_runtime_test_support

tblah reviewed (approved):
Thanks!

kkwli commented on flang/test/lit.cfg.py:
```diff
@@ -203,6 +203,15 @@
 if config.have_openmp_rtl:
     config.available_features.add("openmp_runtime")
+    # For the enabled OpenMP tests, add a substitution that is needed in the tests to find
+    # the omp_lib.{h,mod} files, depending on whether the OpenMP runtime was build as a
```
build → built

mjklemm committed:
Fix typo
